File: kotlin/dfs&bfs/main_7569.kt
fun main() {
    val (m, n, h) = readLine()!!.split(' ').map { it.toInt() }
    val array = Array(h) { Array(n) { readLine()!!.split(' ').map { it.toInt() }.toIntArray() } }
    val deq = ArrayDeque<Pair<Triple<Int, Int, Int>, Int>>()

    var tomato = m * n * h
    var day = 0
    var count = 0
    for (i in 0 until h) {
        for (j in 0 until n) {
            for (k in 0 until m) {
                when (array[i][j][k]) {
                    1 -> deq.add(Triple(i, j, k) to 0)
                    -1 -> tomato--
                }
            }

        }
    }
    while (deq.isNotEmpty()) {
        val pos = deq.removeFirst()
        val (x, y, z) = pos.first
        val level = pos.second
        day = level
        count++
        if (array.getOrNull(x + 1)?.get(y)?.get(z) == 0) {
            deq.add(Triple(x + 1, y, z) to level + 1)
            array[x + 1][y][z] = 1
        }
        if (array.getOrNull(x - 1)?.get(y)?.get(z) == 0) {
            deq.add(Triple(x - 1, y, z) to level + 1)
            array[x - 1][y][z] = 1
        }
        if (array[x].getOrNull(y + 1)?.get(z) == 0) {
            deq.add(Triple(x, y + 1, z) to level + 1)
            array[x][y + 1][z] = 1
        }
        if (array[x].getOrNull(y - 1)?.get(z) == 0) {
            deq.add(Triple(x, y - 1, z) to level + 1)
            array[x][y - 1][z] = 1
        }
        if (array[x][y].getOrNull(z + 1) == 0) {
            deq.add(Triple(x, y, z + 1) to level + 1)
            array[x][y][z + 1] = 1
        }
        if (array[x][y].getOrNull(z - 1) == 0) {
            deq.add(Triple(x, y, z - 1) to level + 1)
            array[x][y][z - 1] = 1
        }
    }
    if (count == tomato)
        println(day)
    else
        println(-1)
}

File: cpp/dfs&bfs/main_7569.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <tuple>
using namespace std;

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	cout.tie(nullptr);

	int m, n, h;
	cin >> m >> n >> h;
	int tomato = m * n * h;
	int*** map = new int**[h];
	deque<pair<tuple<int, int, int>, int>> deq;
	for (int i = 0; i < h; ++i) {
		map[i] = new int*[n];
		for (int j = 0; j < n; ++j) {
			map[i][j] = new int[m];
			for (int k = 0; k < m; ++k) {
				cin >> map[i][j][k];
				if (map[i][j][k] == 1) {
					deq.push_back(make_pair(make_tuple(i, j, k), 0));
				} else if (map[i][j][k] == -1) {
					tomato--;
				}
			}
		}
	}

	int level = 0;
	int count = 0;
	while (!deq.empty()) {
		const tuple<int, int, int> pos = deq.front().first;
		level = deq.front().second;
		deq.pop_front();
		count++;
		int x = get<0>(pos);
		int y = get<1>(pos);
		int z = get<2>(pos);

		if (x > 0 && map[x - 1][y][z] == 0) {
			map[x - 1][y][z] = 1;
			deq.push_back(make_pair(make_tuple(x - 1, y, z), level + 1));
		}
		if (x < h - 1 && map[x + 1][y][z] == 0) {
			map[x + 1][y][z] = 1;
			deq.push_back(make_pair(make_tuple(x + 1, y, z), level + 1));
		}
		if (y > 0 && map[x][y - 1][z] == 0) {
			map[x][y - 1][z] = 1;
			deq.push_back(make_pair(make_tuple(x, y - 1, z), level + 1));
		}
		if (y < n - 1 && map[x][y + 1][z] == 0) {
			map[x][y + 1][z] = 1;
			deq.push_back(make_pair(make_tuple(x, y + 1, z), level + 1));
		}
		if (z > 0 && map[x][y][z - 1] == 0) {
			map[x][y][z - 1] = 1;
			deq.push_back(make_pair(make_tuple(x, y, z - 1), level + 1));
		}
		if (z < m - 1 && map[x][y][z + 1] == 0) {
			map[x][y][z + 1] = 1;
			deq.push_back(make_pair(make_tuple(x, y, z + 1), level + 1));
		}
	}
	if (tomato == count) {
		cout << level;
	} else {
		cout << -1;
	}

}
